galasa-parent/dev.galasa.framework.api.webui/src/main/java/dev/galasa/framework/api/webui/WorklistQuery.java

```diff
@@ -0,0 +1,441 @@
+			
+			IRunResult run = null;
+			for (String runId : runIdsInWorklist) {
+				try {
+					run = getRunById(runId);
+				} catch (ResultArchiveStoreException e) {
+					throw new ServletException("Error while retrieving run for Run ID " + runId, e);
+				}
+				
+				if (run != null) {
+					
+					try {
+						
+						String runName = run.getTestStructure().getRunName();
+						String shortName = run.getTestStructure().getTestShortName();
+						String testClass = run.getTestStructure().getTestName();
+						String result = run.getTestStructure().getResult();
+						
+						WorklistItem worklistItem = new WorklistItem(runId, runName, shortName, testClass, result);
+						
+						runs.add(worklistItem);
+						
+					} catch (ResultArchiveStoreException e) {
+						throw new ServletException("Error while retriving Worklist information for Run ID " + runId, e);
+					}
+					
+				} else {
+					runIdsNotFound.add(runId);
+					try {
+						removeRunIdFromDss(principalUsername, runId);
+					} catch (Exception e) {
+						throw new ServletException("Error while updating the DSS", e);
+					}
+				}
+				
+			}
+			
+			runIdsInWorklist.removeAll(runIdsNotFound);
+		}
+		
+		return runs;
+	}
+	
+	
+	private IRunResult getRunById(String id) throws ResultArchiveStoreException {
+
+		IRunResult run = null;
+	       
+        for (IResultArchiveStoreDirectoryService directoryService : framework.getResultArchiveStore().getDirectoryServices()) {
+           
+           run = directoryService.getRunById(id);
+           
+           if(run != null) {
+        	   return run;
+           }
+           
+        }
+		
+		return null;
+	}
+	
+	
+	private boolean addRunIdToDss(String principalUsername, String runId) throws DynamicStatusStoreException, ServletException {
+		
+		IDynamicStatusStoreService dss = framework.getDynamicStatusStoreService("webui");
+		
+		boolean exceptionThrown = false;
+		
+		do {
+			
+			String oldValue = dss.get("user." + principalUsername + ".worklist");
+			
+			if (oldValue != null) {
+				
+				JsonObject jsonObject = new JsonParser().parse(oldValue).getAsJsonObject();
+				JsonArray runIds = (JsonArray) jsonObject.get("runIds");
+				
+				for (JsonElement el : runIds) {
+					if (el.getAsString().equals(runId)) {
+						// If Run ID is already in Worklist, exit this method
+						return false;
+					}
+				}
+				
+				runIds.add(runId);			
+				
+				String newValue = jsonObject.toString();
+				
+				
+				int count = 0;
+				
+
+				try {
+					dss.performActions(
+							new DssSwap("user." + principalUsername + ".worklist", oldValue, newValue));
+					exceptionThrown = false;
+				} catch (DynamicStatusStoreException e) {
+					exceptionThrown = true;
+					oldValue = dss.get("user." + principalUsername + ".worklist");
+					count++;
+					
+					if (count > 1000) {
+						throw new ServletException("Error while updating the DSS, maximum attempts reached");
+					}
+				}
+					
+				return false;
```

Comment:
this should not be here,  as will cause only one pass through the loop,    did you test your changes?